server/src/com/thoughtworks/go/server/web/BackupFilter.java

```diff
@@ -0,0 +1,131 @@
+    }
+
+    private String urlEncode(String string) throws UnsupportedEncodingException {
+        return URLEncoder.encode(string, "UTF-8");
+    }
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+        this.filterConfig = filterConfig;
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        SpringBeanAutowiringSupport.processInjectionBasedOnCurrentContext(this);
+
+        String url = ((HttpServletRequest) request).getRequestURI();
+        if (isBackupFinishJsonUrl(url)) {
+            response.setContentType("application/json");
+            JsonObject json = new JsonObject();
+            json.addProperty("is_backup_finished", backupService.isBackingUp());
+            response.getWriter().print(json);
+            return;
+        }
+        if (backupService.isBackingUp()) {
+            ((HttpServletResponse) response).setHeader("Cache-Control", "private, max-age=0, no-cache");
+            ((HttpServletResponse) response).setDateHeader("Expires", 0);
+
+            if (isAPIUrl(url) && !isMessagesJson(url)) {
+                generateAPIResponse(request, response);
+            } else {
+                try {
+                    generateHTMLResponse(response);
+
+                } catch (Exception e) {
+                    e.printStackTrace();
+                }
+            }
+            return;
+        } else {
+            chain.doFilter(request, response);
+        }
+
+    }
+
+    private void generateHTMLResponse(ServletResponse response) throws IOException {
+        InputStream resourceAsStream = filterConfig.getServletContext().getResourceAsStream("/static/backup_in_progress.html");
+        response.setContentType("text/html");
+        StringWriter writer = new StringWriter();
+        IOUtils.copy(resourceAsStream, writer);
+        String content = writer.toString();
+        content = content.replaceAll("%backup_initiated_by%", urlEncode(backupService.backupRunningSinceISO8601()));
+        content = content.replaceAll("%backup_started_by%", urlEncode((backupService.backupStartedBy())));
+        response.getWriter().print(content);
+    }
+
+    private void generateAPIResponse(ServletRequest request, ServletResponse response) throws IOException {
+        response.setContentType(request.getContentType());
+        JsonObject jsonObject = new JsonObject();
+        jsonObject.addProperty("message", "Server is under maintenance mode, please try later.");
+        response.getWriter().print(jsonObject);
+        ((HttpServletResponse) response).setStatus(503);
+    }
+
+    private boolean isBackupFinishJsonUrl(String url) {
+        return "/go/is_backup_finished.json".equals(url);
+    }
+
+    private boolean isAPIUrl(String url) {
+        Pattern pattern = Pattern.compile("^((.*/api/.*)|(.*[^/]+\\.(xml|json)(\\?.*)?))$");
```

Comment:
Compiling patterns may be an expensive operation to perform on each request. I'd consider making this a constant.